addons/spreadsheet/static/src/global_filters/helpers.js

```diff
@@ -36,6 +36,47 @@ export const RELATIVE_PERIODS = {
     year_to_date: _t("Year to Date"),
 };
 
+/**
+ * @param {DateValue} dateFilterValue
+ * @param {{ chain: string, type: string }} [fieldMatching]
+ * @returns {string}
+ */
+export function getBestGranularity(dateFilterValue, fieldMatching) {
+    if (!dateFilterValue) {
+        return "year";
+    }
+    const { from, to } = getDateRange(dateFilterValue);
+    const numberOfDays = to.diff(from, "days").days;
+    if (numberOfDays <= 1) {
+        return fieldMatching?.type === "datetime" ? "hour" : "day";
+    } else if (numberOfDays <= 90) {
+        return "day";
+    } else if (numberOfDays <= 365 * 3) {
+        return "month";
+    } else {
+        return "year";
+    }
+}
+
+/**
+ * @param {DateValue} dateFilterValue
+ * @returns {string[]}
+ */
+export function getValidGranularities(dateFilterValue) {
+    if (!dateFilterValue) {
+        return ["month", "quarter", "year"];
+    }
+    const { from, to } = getDateRange(dateFilterValue);
+    const numberOfDays = to.diff(from, "days").days;
+    if (numberOfDays <= 31) {
+        return ["hour", "day"];
+    } else if (numberOfDays <= 31 * 3) {
+        return ["day", "week", "month"];
+    } else {
+        return ["month", "quarter", "year"];
+    }
```

Comment:
So we return "year" if the chart has 4 months within the same year ? That doesn't seems right ...